core/src/main/java/org/mule/runtime/core/privileged/processor/MessageProcessors.java

```diff
@@ -307,35 +380,45 @@ public static BaseEventContext newChildContext(CoreEvent event, Optional<Compone
         })
         .transform(processor)
         .doOnNext(completeSuccessIfNeeded())
-        .mergeWith(Flux.<CoreEvent>create(errorSwitchSinkSinkRef))
+        .map(event -> right(MessagingException.class, event))
+        // This Either here is used to propagate errors. If the error is sent directly through the merged with Flux, it will be
+        // cancelled, ignoring the onErrorcontinue of the parent Flux.
+        .mergeWith(Flux.<Either<MessagingException, CoreEvent>>create(errorSwitchSinkSinkRef))
+        .map(result -> result.reduce(me -> {
+          throw propagateWrappingFatal(me);
+        }, response -> response))
         .distinct(event -> (BaseEventContext) event.getContext(), () -> seenContexts)
-        .map(result -> quickCopy((((BaseEventContext) result.getContext()).getParentContext().get()), result));
+        .map(result -> toParentContext(result));
   }
 
   private static void childContextResponseHandler(CoreEvent eventChildCtx,
-                                                  SinkRecorderToReactorSinkAdapter<CoreEvent> errorSwitchSinkSinkRef,
+                                                  SinkRecorderToReactorSinkAdapter<Either<MessagingException, CoreEvent>> errorSwitchSinkSinkRef,
                                                   boolean completeParentIfEmpty) {
     Mono.from(((BaseEventContext) eventChildCtx.getContext()).getResponsePublisher())
         .doOnSuccess(response -> {
           if (response == null && completeParentIfEmpty) {
             ((BaseEventContext) eventChildCtx.getContext()).getParentContext().get().success();
             errorSwitchSinkSinkRef.next();
           } else {
-            errorSwitchSinkSinkRef.next(response);
+            if (response == null) {
+              errorSwitchSinkSinkRef.next();
+            } else {
+              errorSwitchSinkSinkRef.next(right(MessagingException.class, response));
+            }
           }
         })
         .onErrorResume(MessagingException.class, error -> {
-          final CoreEvent event = error.getEvent();
-          errorSwitchSinkSinkRef
-              .error(new MessagingException(quickCopy(((BaseEventContext) event.getContext()).getParentContext().get(), event),
-                                            error));
-
+          errorSwitchSinkSinkRef.next(left(new MessagingException(toParentContext(error.getEvent()), error), CoreEvent.class));
           return Mono.empty();
         })
         .doOnError(e -> LOGGER.error("Uncaugh exception in childContextResponseHandler", e))
         .toProcessor();
   }
 
+  private static CoreEvent toParentContext(final CoreEvent event) {
+    return quickCopy(((BaseEventContext) event.getContext()).getParentContext().get(), event);
```

Comment:
![Bug](https://sonarsource.github.io/sonar-github/bug.png 'Bug') Bug: Call "Optional#isPresent()" before accessing the value. ([squid:S3655](https://sonarqube.kbuild.msap.io/coding_rules#rule_key=squid%3AS3655))

[See it in SonarQube](https://sonarqube.kbuild.msap.io/project/issues?id=mule-runtime-mule&pullRequest=7683&issues=AWjyMA29otIXnmsUSXGc&open=AWjyMA29otIXnmsUSXGc)

[ISSUE_KEY]: AWjyMA29otIXnmsUSXGc